text/3735-field-projections.md

```diff
@@ -0,0 +1,1344 @@
+
+So for example:
+
+```rust
+#[projecting]
+#[repr(transparent)]
+pub struct Container<T> {
+    inner: T,
+}
+
+struct Foo {
+    bar: i32,
+}
+```
+
+Now `Container<Foo>` has a [field type] associated with `bar` implementing `Field` with:
+- `Base = Container<Foo>`
+- `Type = Container<i32>`
+- `OFFSET = offset_of!(Foo, bar)`
+
+The same is true for the non-generic case:
+
+```rust
+#[projecting]
+#[repr(transparent)]
+pub struct Container {
+    inner: Foo,
+}
+
+struct Foo {
+    bar: i32,
+}
+```
+
+Here are some error examples:
+
+```rust
+// ERROR: missing `#[repr(transparent)]`
+#[projecting]
+pub struct Container<T> {
+    inner: T,
+}
+
+// ERROR: no field to project onto found, the struct has no fields
+#[projecting]
+#[repr(transparent)]
+pub struct Container {}
+
+// ERROR: no field to project onto found, all fields are always zero-sized
+#[projecting]
+#[repr(transparent)]
+pub struct Container<T> {
+    phantom1: PhantomData<fn() -> T>,
+    phantom2: PhantomData<fn(T)>,
+}
+```
+
+### Field Projection Operator
+
+The field projection operator `->` has the following syntax:
+
+> **<sup>Syntax</sup>**\
+> _ProjectionExpression_ :\
+> &nbsp;&nbsp; [_Expression_] `->` _ProjectionMember_
+>
+> _ProjectionMember_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; [IDENTIFIER]
+> &nbsp;&nbsp; | [TUPLE_INDEX]
+
+[IDENTIFIER]: https://doc.rust-lang.org/reference/identifiers.html
+[_Expression_]: https://doc.rust-lang.org/reference/expressions.html
+[TUPLE_INDEX]: https://doc.rust-lang.org/reference/tokens.html#tuple-index
+
+#### Desugaring
+
+```rust
+struct T {
+    field: F,
+}
+
+let t: C<T> = /* ... */;
+let _ = t->field;
+
+// becomes
+
+let _ = Project::<field_of!(<C<T> as Projectable>::Inner, field)>::project(t);
+```
+
+The `C<T>` in the `C<T> as Projectable` comes from a type inference variable over the expression
+`t`.
+
+
+#### `Projectable` & `Project` Traits
+
+Added to `core::ops`:
+
+```rust
+pub trait Projectable: Sized {
+    type Inner: ?Sized;
+}
+
+pub trait Project<F>: Projectable
+where
+    F: Field<Base = Self::Inner>,
+{
+    type Output;
+
+    fn project(self) -> Self::Output;
+}
+```
+
+## Stdlib Field Projections
+
+All examples from the guide-level explanation work when the standard library is extended with the
+implementations detailed below.
+
+The following pointer types get an implementation for `Projectable` with `Inner = T`. They support
+projections for any field and perform the obvious offset operation.
+
+- `&mut T`
+- `&T`
+- `*mut T`
+- `*const T`
+- `NonNull<T>`
+- `Cow<'_, T>`
+
+The following types get annotated with [`#[projecting]`](#projecting-attribute):
+
+- `MaybeUninit<T>`
+- `Cell<T>`
+- `UnsafeCell<T>`
+- `SyncUnsafeCell<T>`
+
+### Pin Projections
+
+In order to provide [pin projections], a new derive macro `PinProject` and a trait `PinField` is
+required:
+
+```rust
+pub unsafe trait PinField: Field {
+    type Projected<'a>;
+
+    fn from_pinned_ref<'a>(r: &'a mut Self::Type) -> Self::Projected<'a>;
```

Comment:
```suggestion
    unsafe fn from_pinned_ref<'a>(r: &'a mut Self::Type) -> Self::Projected<'a>;
```
This method should be unsafe because it can cause UB if called incorrectly. As defined right now safe code can call it. In fact I'm not sure the trait itself needs to be `unsafe` since the only safety requireemnt here is calling the method correctly.

Reply:
> how about changing the signature to this

I believe this is incorrect in the non-pinned case: that method would convert a `Pin<&mut Field>` to a `&mut Field` and expose that `&mut Field` to user code, who could then swap the value out, thus breaking the `Pin` invariant.

> I still would prefer the trait to be unsafe, in order to ensure that this always represents a valid field projection for other unsafe code to rely upon

oh yeah that's important, agreed

Reply:
fixed